### PR TITLE
feat(sequencer): conditional unbonding + kick for liveness

### PR DESCRIPTION
**Pull request by danwt**
#### features:

- reworked x/sequencer and slashing according to adr https://www.notion.so/dymension/Sequencer-Bond-Management-Slashing-117a4a51f86a807998d8e3213ae3c1f2
- added extra bond conditions in rollapp keeper and light client keeper according to https://www.notion.so/dymension/ADR-x-Sequencer-Unbonding-Conditioned-by-IBC-LightClient-Updates-and-Rollapp-State-Updates-120a4a51f86a808f9dece8139ee3bd30

#### changes needed:

- [ ] x/rollapp
    - [ ] unbond cond rollapp state infos
        - [x] prune
        - [x] bookkeeping
        - [x] condition
        - [x] test prune
        - [x] test bookkeeping
        - [x] test condition
        - [x] wiring to app
    - [x] liveness
        - [x] logic
        - [x] test
        - [x] params
        - [x] params validation
        - [x] params validation test
    - [x] kick hook
        - [x] logic
        - [ ] test
    - [x] proposer change hook
        - [x] logic
        - [ ] test
- [ ] x/sequencer
    - [x] kick proposer
        - [x] msg
        - [x] validations
        - [x] logic
        - [x] test
        - [x] cli
        - [x] hook to rollapp keeper
        - [x] hook wiring
        - [x] param
        - [x] param test
        - [x] flow test
        - [x] event
    - [x] update opt in status
        - [x] msg
        - [x] validations
        - [x] logic
        - [x] test
        - [x] cli
        - [x] event
    - [ ] genesis import export
        - [x] logic
        - [x] test
    - [x] proposer change hook
        - [x] logic
        - [ ] test
    - [x] events
        - [x] rotation complete
        - [x] rotation start
        - [x] notice start
        - [x] unbonded
        - [x] bond increase
        - [x] create seq
        - [x] proposer change
        - [x] slash
    - [ ] audit
        - [ ] unsafe coins usage
        - [ ] correct address type usage
    - [ ] invariants
        - [ ] new ones added
    - [ ] bond change flow
        - [x] logic
        - [x] test happy
        - [ ] test ugly
    - [ ] rotation flow
         - [x] logic
         - [x] test happy
         - [ ] test ugly
    - [ ] slash flow
        - [x] logic
        - [x] test happy
        - [ ] test ugly
        - [x] extra test cases for in conjunction with rotation
    - [ ] params
        - [x] params validation
        - [x] params validation test
- [ ] x/lightclient
    - [ ] unbond cond light client states
        - [x] bookkeeping
        - [x] prune
        - [x] condition
        - [x] test prune
        - [x] test bookkeeping
        - [x] test condition
        - [x] wiring to app
        - [x] delete old bookkeeping code
    - [x] unbonding period removal
    - [x] enabled/disable flag
    - [x] genesis import export
        - [x] logic
        - [x] test
        
#### tech debts and todos:

- [ ] https://github.com/dymensionxyz/dymension/issues/1357
- [ ] https://github.com/dymensionxyz/dymension/issues/1360
- [ ] https://github.com/dymensionxyz/dymension/issues/1364
- [ ] https://github.com/dymensionxyz/dymension/issues/1365
- [ ] https://github.com/dymensionxyz/dymension/issues/1366

#### remarks:

- Have avoided changing events and TX/Query where possible to avoid changing frontends, which leaves things less pretty 
- Should really have more tests
- Added an ugly test enable/disable for lightclient to not break ibctesting/

----

Closes https://github.com/dymensionxyz/dymension/issues/1328

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

----

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: x/sequencer/keeper/msg_server_bond.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	defer func() {
+		k.SetSequencer(ctx, seq)
+	}()
+
+	seq.OptedIn = false // ensures they will not get chosen as their own successor!
+	err = k.TryUnbond(ctx, &seq, seq.TokensCoin())
+	if errorsmod.IsOf(err, types.ErrUnbondProposerOrSuccessor) {
+		// not allowed to unbond immediately, need to serve a notice to allow the rollapp community to organise
+
```
Comment (omritoptix) [inline x/sequencer/keeper/msg_server_bond.go:73]:
```suggestion
// we wanna make sure that if the proposer notice period elapsed, we're not waiting for his last block still. 
// otherwise we don't want to let him unbond.
```

File: x/sequencer/keeper/proposer.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	seqs := k.RollappPotentialProposers(ctx, rollapp)
+	successor = ProposerChoiceAlgo(seqs)
+	k.SetSuccessor(ctx, rollapp, successor.Address)
+}
+
+// ProposerChoiceAlgo : choose the one with most bond
+// Requires sentinel to be passed in, as last resort.
+func ProposerChoiceAlgo(seqs []types.Sequencer) types.Sequencer {
+	if len(seqs) == 0 {
```
Comment (omritoptix) [inline x/sequencer/keeper/proposer.go:99]: panicing here will halt the chain

**Review by omritoptix — COMMENTED** [review]
great job!

File: x/rollapp/keeper/hooks_listeners.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	sequencertypes "github.com/dymensionxyz/dymension/v3/x/sequencer/types"
+)
+
+var _ sequencertypes.Hooks = SequencerHooks{}
+
+type SequencerHooks struct {
+	*Keeper
+}
+
+func (k SequencerHooks) AfterChooseNewProposer(ctx sdk.Context, rollapp string, before, after sequencertypes.Sequencer) {
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/rollapp/keeper/hooks_listeners.go:14]: Yeah right! There is a tech debt to rename this field
I didn't want to touch the rollapp protos

File: x/lightclient/types/params.go [hunk 4]
```diff
@@ -11,11 +11,14 @@ import (
 	commitmenttypes "github.com/cosmos/ibc-go/v7/modules/core/23-commitment/types"
 	ibctm "github.com/cosmos/ibc-go/v7/modules/light-clients/07-tendermint"
 	ics23 "github.com/cosmos/ics23/go"
-	sequencertypes "github.com/dymensionxyz/dymension/v3/x/sequencer/types"
 )
 
 func DefaultExpectedCanonicalClientParams() ibctm.ClientState {
-	return ExpectedCanonicalClientParams(sequencertypes.DefaultUnbondingTime)
+	// Note: need to be very sure that this is the same value that the
+	// relaye gets when it queries the rollapp (x/sequencers)
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/lightclient/types/params.go:18]: the relayer is taking it from the rollapp not the hub
I will do a better comment

File: x/rollapp/keeper/hooks_listeners.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (danwt) [reply x/rollapp/keeper/hooks_listeners.go:14]: https://github.com/dymensionxyz/dymension/issues/1357
It was a separate issue but I rolled it into this

File: x/sequencer/keeper/msg_server_bond.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		if !seq.NoticeInProgress(ctx.BlockTime()) {
+			if !(k.IsProposer(ctx, seq) && k.awaitingLastProposerBlock(ctx, seq.RollappId)) {
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/sequencer/keeper/msg_server_bond.go:75]: Not sure I got you. I tried to keep it exactly the same as it already is.

Can you rephrase?

[the remark this answers is not in the document]
Comment (omritoptix) [reply x/sequencer/keeper/msg_server_bond.go:75]: yea sorry. seems like this function doesn't restrict a successor from starting his notice period. which afair we don't want to allow a successor to be able to start his notice period (i.e he needs first to become a proposer, and only after that).

hope it's clearer.

[the remark this answers is not in the document]
Comment (danwt) [reply x/sequencer/keeper/msg_server_bond.go:75]: Ah I see

>afair we don't want to allow a successor to be able to start his notice period (i.e he needs first to become a proposer, and only after that).

I'm fine with that, it's much simpler actually! But it's not how it currently works on main FYI

File: x/sequencer/keeper/msg_server_bond.go (hunk 1 above)
Comment (danwt) [reply x/sequencer/keeper/msg_server_bond.go:73]: I added something similar

File: x/sequencer/keeper/msg_server_bond.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (danwt) [reply x/sequencer/keeper/msg_server_bond.go]: Implemented

File: x/sequencer/keeper/proposer.go (hunk 2 above)
Comment (danwt) [reply x/sequencer/keeper/proposer.go:106]: fixed to propogate an error
(although in this case I'm not sure it's a net benefit)

**Review by omritoptix — APPROVED** [review]

File: x/lightclient/keeper/keeper.go [hunk 6]
```diff
@@ -1,28 +1,55 @@
...
+
+// wrapper to allow taking a pointer to mutable value
+type enabled struct {
+	enabled bool
+}
+
 type Keeper struct {
+	// if false, will not run the msg update client ante handler. Very hacky
+	// use to avoid problems in ibctesting.
+	enabled *enabled
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/lightclient/keeper/keeper.go:33]: I'll try

File: ibctesting/light_client_test.go
[the remark this answers is not in the document]
Comment (danwt) [reply ibctesting/light_client_test.go:1]: It just doesn't work. The problem is the ibctesting is not doing a realistic rollapp setup. It creates the light client, then creates the rollapp, which is not how it works in reality.

I spoke with @omritoptix about it - just don't have time to rework the whole of ibctesting now. But yeah for sure need to test better.There is a tech debt issue https://github.com/dymensionxyz/dymension/issues/1364

File: x/lightclient/ante/ibc_msg_update_client.go [hunk 7]
```diff
@@ -1,82 +1,153 @@
...
+	if stateInfos.containingHPlus1 != nil {
+		// the header is pessimistic: the state update has already been received, so we check the header doesn't mismatch
+		return errorsmod.Wrap(i.validateUpdatePessimistically(ctx, stateInfos, header.ConsensusState(), h), "validate pessimistic")
+	}
+
+	// the header is optimistic: the state update has not yet been received, so we save optimistically
+	return errorsmod.Wrap(i.k.SaveSigner(ctx, seq.Address, msg.ClientId, h), "save updater")
+}
+
+var (
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/lightclient/ante/ibc_msg_update_client.go:81]: the errors you mean?
I prefer not since they are just internal/local to the methods in this file

File: x/lightclient/keeper/keeper.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (danwt) [reply x/lightclient/keeper/keeper.go:33]: It's also needed to be accessed by x/rollapp so hard to remove tbh
There is a tech debt https://github.com/dymensionxyz/dymension/issues/1365

File: ibctesting/light_client_test.go
[the remark this answers is not in the document]
Comment (mtsitrin) [reply ibctesting/light_client_test.go:1]: better to leave the tests and have `t.SkipNow()`

File: x/lightclient/ante/ibc_msg_update_client.go [hunk 8]
```diff
@@ -1,82 +1,153 @@
...
 	if err != nil {
-		// No BDs found for next height.
-		// Will accept the update optimistically
-		// But also save the blockProposer address with the height for future verification
-		i.acceptUpdateOptimistically(ctx, msg.ClientId, header)
-		return nil
+		return stateInfos{}, err
+	}
+	s1 := s0
+	if !s1.ContainsHeight(h) {
```
Comment (mtsitrin) [inline x/lightclient/ante/ibc_msg_update_client.go:130]: how it's possbile?
if h not found in `s0`, it means it's future height
how `h+1` can be found?
maybe u meant `if !s0.ContainsHeight(h+1) {`?

File: x/lightclient/ante/ibc_msg_update_client.go [hunk 9]
```diff
@@ -1,82 +1,153 @@
...
+		return errorsmod.Wrap(err, "get state infos")
+	}
+
+	if stateInfos.containingHPlus1 != nil {
+		// the header is pessimistic: the state update has already been received, so we check the header doesn't mismatch
+		return errorsmod.Wrap(i.validateUpdatePessimistically(ctx, stateInfos, header.ConsensusState(), h), "validate pessimistic")
+	}
+
+	// the header is optimistic: the state update has not yet been received, so we save optimistically
+	return errorsmod.Wrap(i.k.SaveSigner(ctx, seq.Address, msg.ClientId, h), "save updater")
```
Comment (mtsitrin) [inline x/lightclient/ante/ibc_msg_update_client.go:78]: this wrapping of possible `nil` err is not intuitive
I would suggest to stick with convention (err check + wrap on error != nil)

File: x/lightclient/ante/ibc_msg_update_client.go [hunk 10]
```diff
@@ -1,82 +1,153 @@
...
-	sequencerPubKey, err := i.lightClientKeeper.GetSequencerPubKey(ctx, stateInfo.Sequencer)
+	return stateInfos{s0, s1}, nil
+}
+
+func (i IBCMessagesDecorator) validateUpdatePessimistically(ctx sdk.Context, infos stateInfos, consState *ibctm.ConsensusState, h uint64) error {
+	bd, _ := infos.containingH.GetBlockDescriptor(h)
+	seq, err := i.k.SeqK.RealSequencer(ctx, infos.containingHPlus1.Sequencer)
 	if err != nil {
-		return err
+		return gerrc.ErrInternal.Wrap("get sequencer of state info")
```
Comment (mtsitrin) [inline x/lightclient/ante/ibc_msg_update_client.go:146]: missing the original err

File: x/lightclient/ante/ibc_msg_update_client.go [hunk 11]
```diff
@@ -1,82 +1,153 @@
...
-		return nil
+		return nil, errNoHeader
 	}
+	return header, nil
+}
+
+// if containingHPlus1 is not nil then containingH also guaranteed to not be nil
+type stateInfos struct {
+	containingH      *rollapptypes.StateInfo
+	containingHPlus1 *rollapptypes.StateInfo
```
Comment (mtsitrin) [inline x/lightclient/ante/ibc_msg_update_client.go:116]: `containingHPlus1` used to get next sequencer
can we take the `NextSequencer` field from the `containingH` state info instead? (after the HF PR)

File: x/lightclient/types/state.go [hunk 12]
```diff
@@ -24,11 +22,11 @@ func CheckCompatibility(ibcState ibctm.ConsensusState, raState RollappState) err
 		return errorsmod.Wrap(ErrStateRootsMismatch, "block descriptor state root does not match tendermint header app hash")
 	}
 	// Check if the nextValidatorHash matches for the sequencer for h+1 block descriptor
-	nextValHashFromStateInfo, err := GetValHashForSequencer(raState.NextBlockSequencer)
+	hash, err := raState.NextBlockSequencer.ValsetHash()
 	if err != nil {
-		return errors.Join(ErrValidatorHashMismatch, err)
+		return gerrc.ErrInternal.Wrap("val set hash")
```
Comment (mtsitrin) [inline x/lightclient/types/state.go:27]: missing original err

File: x/lightclient/keeper/hook_listener.go [hunk 13]
```diff
@@ -31,75 +35,114 @@ func (hook rollappHook) AfterUpdateState(
...
-		// Check if any optimistic updates were made for the given height
-		blockValHash, found := hook.k.GetConsensusStateValHash(ctx, canonicalClient, bd.GetHeight())
-		if !found {
-			continue
-		}
-		err := hook.checkStateForHeight(ctx, rollappId, bd, canonicalClient, sequencerPk, blockValHash)
-		if err != nil {
-			return err
-		}
+		return gerrc.ErrInternal.Wrap("get sequencer for state info")
```
Comment (mtsitrin) [inline x/lightclient/keeper/hook_listener.go:53]: missing original err

File: x/lightclient/keeper/hook_listener.go [hunk 14]
```diff
@@ -31,75 +35,114 @@ func (hook rollappHook) AfterUpdateState(
...
-		if err != nil {
-			return err
-		}
-		bd, _ := previousStateInfo.GetBlockDescriptor(stateInfo.StartHeight - 1)
-		err = hook.checkStateForHeight(ctx, rollappId, bd, canonicalClient, sequencerPk, blockValHash)
-		if err != nil {
-			return err
+
+	// [h-1..,h) is correct because we compare against a next validators hash
+	for h := stateInfo.GetLatestHeight() - 1; h < stateInfo.GetLatestHeight(); h++ {
```
Comment (mtsitrin) [inline x/lightclient/keeper/hook_listener.go:57]: not sure why u need for loop. seems it will run single time

File: x/sequencer/keeper/get_and_set.go [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+		ret = append(ret, seq)
+	}
+	return ret
+}
+
+func (k Keeper) GetProposer(ctx sdk.Context, rollapp string) types.Sequencer {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.ProposerByRollappKey(rollapp))
+	if bz == nil {
+		return k.GetSequencer(ctx, types.SentinelSeqAddr)
```
Comment (mtsitrin) [inline x/sequencer/keeper/get_and_set.go:154]: why not return `k.SentinelSequencer(ctx)`

File: x/sequencer/keeper/get_and_set.go [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	return k.GetSequencer(ctx, string(bz))
+}
+
+func (k Keeper) GetSuccessor(ctx sdk.Context, rollapp string) types.Sequencer {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.SuccessorByRollappKey(rollapp))
+	if bz == nil {
+		return k.GetSequencer(ctx, types.SentinelSeqAddr)
```
Comment (mtsitrin) [inline x/sequencer/keeper/get_and_set.go:163]: how u distinguish between No successer set (no rotation in progress), and sentinal successor?

File: x/sequencer/keeper/rotation.go [hunk 17]
```diff
@@ -12,135 +11,82 @@ import (
...
-			sdk.NewAttribute(types.AttributeKeyRollappId, rollappId),
-			sdk.NewAttribute(types.AttributeKeyNextProposer, nextProposer.Address),
-			sdk.NewAttribute(types.AttributeKeyRewardAddr, nextProposer.RewardAddr),
-			sdk.NewAttribute(types.AttributeKeyWhitelistedRelayers, strings.Join(nextProposer.WhitelistedRelayers, ",")),
-		),
-	)
+func (k Keeper) awaitingLastProposerBlock(ctx sdk.Context, rollapp string) bool {
+	proposer := k.GetProposer(ctx, rollapp)
+	return proposer.NoticeElapsed(ctx.BlockTime())
 }
```
Comment (mtsitrin) [inline x/sequencer/keeper/rotation.go:70]: we discussed in the past that this is not reliable enough.
`awaitingLastProposerBlock` should be true only after `chooseSuccessor` is called

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: ibctesting/light_client_test.go
[the remark this answers is not in the document]
Comment (danwt) [reply ibctesting/light_client_test.go:1]: I disagree, the code is in git.

File: x/lightclient/ante/ibc_msg_update_client.go [hunk 18]
```diff
@@ -1,82 +1,153 @@
...
+	}
+	seq, err := i.getSequencer(ctx, header)
+	err = errorsmod.Wrap(err, "get sequencer")
+	if errorsmod.IsOf(err, errProposerMismatch) {
+		// this should not occur on any chain, regardless of being a rollapp or not
+		return err
+	}
+	if errorsmod.IsOf(err, gerrc.ErrNotFound) {
+		if !canonical {
+			// not from sequencer, and not canonical - it's not interesting
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/lightclient/ante/ibc_msg_update_client.go:42]: What's your suggestion, concretely? the check is saying it's not from sequencer AND it's not canonical

File: x/lightclient/ante/ibc_msg_update_client.go [hunk 19]
```diff
@@ -1,82 +1,153 @@
...
+	}
+	if err != nil {
+		return err
+	}
+
+	// ~~~~~
+	// now we know that the msg is a header, and it was produced by a sequencer
+	// ~~~~~
+
+	if !seq.Bonded() {
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/lightclient/ante/ibc_msg_update_client.go:55]: > can move above, as part of getSequencer's error checking

I'm taking it as a nit, ignoring for now. But 👍 .

>also can validate the seq's rollappId with the expected

What's the benefit?

File: x/lightclient/ante/ibc_msg_update_client.go (hunk 8 above)
Comment (danwt) [reply x/lightclient/ante/ibc_msg_update_client.go:130]: great catch! Thanks

File: x/lightclient/ante/ibc_msg_update_client.go (hunk 9 above)
Comment (danwt) [reply x/lightclient/ante/ibc_msg_update_client.go:78]: errorsmod is specifically designed to allow this

Comment (danwt) [reply x/lightclient/ante/ibc_msg_update_client.go:78]: > // If err is nil, this returns nil, avoiding the need for an if statement when
// wrapping a error returned at the end of a function

File: x/lightclient/ante/ibc_msg_update_client.go (hunk 10 above)
Comment (danwt) [reply x/lightclient/ante/ibc_msg_update_client.go:146]: fixed

File: x/lightclient/ante/ibc_msg_update_client.go (hunk 11 above)
Comment (danwt) [reply x/lightclient/ante/ibc_msg_update_client.go:116]: sure

File: x/lightclient/keeper/hook_listener.go (hunk 13 above)
Comment (danwt) [reply x/lightclient/keeper/hook_listener.go:53]: fixed

File: x/sequencer/genesis.go [hunk 20]
```diff
@@ -6,45 +6,57 @@ import (
...
-			k.AddSequencerToUnbondingQueue(ctx, &elem)
-		} else if elem.IsNoticePeriodInProgress() {
-			k.AddSequencerToNoticePeriodQueue(ctx, &elem)
+		if err := k.SetSequencerByDymintAddr(ctx, elem.MustProposerAddr(), elem.Address); err != nil {
+			panic(err)
 		}
 	}
 
+	for _, s := range genState.NoticeQueue {
+		seq := k.GetSequencer(ctx, s)
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/sequencer/genesis.go:20]: The notice queue will not have sentinel

File: x/sequencer/keeper/rotation.go (hunk 17 above)
Comment (danwt) [reply x/sequencer/keeper/rotation.go:70]: Right, good spot, I forgot to move ChooseSuccessorForFinishedNotices to begin block!

File: x/sequencer/keeper/get_and_set.go (hunk 16 above)
Comment (danwt) [reply x/sequencer/keeper/get_and_set.go:163]: Why do you need to do that?
If you need to do that,  awaitingLastProposerBlock
Can expose in a query if need be

File: x/lightclient/types/state.go (hunk 12 above)
Comment (danwt) [reply x/lightclient/types/state.go]: fixed

File: x/lightclient/ante/ibc_msg_update_client.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (mtsitrin) [reply x/lightclient/ante/ibc_msg_update_client.go:42]: my q is why `AND it's not canonical` required? why not return an error if it's not from a sequencer

[the remark this answers is not in the document]
Comment (danwt) [reply x/lightclient/ante/ibc_msg_update_client.go:42]: if it's not from sequencer then we would reject ibc updates for non rollapps
